### PR TITLE
Fixed #708: Replace HttpClient with OkHttp

### LINKED ISSUE #708: Apache HTTP Client Removal/Deprecated in Android M.
Hi,
We know that in android m the Apache HTTP Client is removed/deprecated.
Workaround is to have this in the gradle.

 
"To continue using the Apache HTTP APIs, you must first declare the following compile-time dependency in your build.gradle file:"
android {
    useLibrary 'org.apache.http.legacy'
}
https://developer.android.com/preview/behavior-changes.html#behavior-apache-http-client

What are the insights the Couchbase team taken for this change from android team.I think "apache.http.client" widely in used "couchbase-lite-java-core".

How will be the future of coucbaselite if its removed natively completely? Use as jar is option and will it work all ok?

http://developer.android.com/sdk/api_diff/23/changes.html

https://github.com/couchbase/couchbase-lite-java-core/search?utf8=%E2%9C%93&q=org.apache.http.client&type=Code

Thanks
Nithin

### PR DESCRIPTION
**Pull request by hideki**
Original Ticket: [Apache HTTP Client Removal/Deprecated in Android M.](https://github.com/couchbase/couchbase-lite-java-core/issues/708)

- Removed HttpClient from couchbase-lite-java-core completely and replaced it with OkHttp
- Rewrite `RemoteRequest` and its all sub-classes. Moved these classes from `support` package to `replicator` package
- Removed all Mime/Multipart extension java classes based on HttpClient. `com.couchbase.org.apache.http.entitity.mime.*`
- Added Copyright for all java files that did not include.

Area needed to be reviewed:
- `com.couchbase.lite.replicator.*`
- `CouchbaseLiteHttpClientFactory`, `MultipartDocumentReader`, `PersistentCookieJar`, `SerializableCookie` and `Utils` in `com.couchbase.lite.support`

### CODE REVIEW

File: src/main/java/com/couchbase/lite/replicator/ChangeTracker.java [hunk 1]
```diff
@@ -295,214 +288,161 @@ protected void runLoop() {
...
-
                     // In case response status is Error, ChangeTracker stops here
-                    if(isResponseFailed(response)) {
-                        if(retryIfFailedPost(response.getStatusLine()))
+                    if (isResponseFailed(response)) {
+                        if (retryIfFailedPost(response)) {
+                            RequestUtils.closeResponseBody(response);
                             continue;
+                        }
+                        RequestUtils.closeResponseBody(response);
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/ChangeTracker.java]: RequestUtils.closeResponseBody(response) in line 326 and 323 can be moved up to before line 322.

File: src/main/java/com/couchbase/lite/replicator/RemoteRequestRetry.java [hunk 2]
```diff
@@ -118,7 +137,7 @@ public CustomFuture submit() {
      * @param gzip true - send gzipped request
      */
     public CustomFuture submit(boolean gzip) {
-        RemoteRequest request = generateRemoteRequest();
+        com.couchbase.lite.replicator.RemoteRequest request = generateRemoteRequest();
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/RemoteRequestRetry.java]: Can't we just have `RemoteRequest`?

File: src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java [hunk 3]
```diff
@@ -1687,22 +1686,14 @@ public void setCookie(String name, String value, String path, long maxAge, boole
...
-            cookie.setPath(remote.getPath());
-        }
-
-        cookie.setExpiryDate(expirationDate);
-        cookie.setSecure(secure);
-        List<Cookie> cookies = Collections.singletonList((Cookie) cookie);
+        if (path == null || path.length() == 0)
+            path = remote.getPath();
+        Cookie.Builder builder = new Cookie.Builder();
+        builder.name(name).value(value).domain(remote.getHost()).path(path).expiresAt(expirationDate.getTime());
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java]: Can expirationDate be null?

File: src/main/java/com/couchbase/lite/support/CouchbaseLiteHttpClientFactory.java [hunk 4]
```diff
@@ -120,62 +96,41 @@ public void deleteCookie(String name) {
...
+            if (!cookie.name().equals(name))
                 retainedCookies.add(cookie);
-            }
-        }
-        cookieStore.clear();
-        for (Cookie retainedCookie : retainedCookies) {
-            cookieStore.addCookie(retainedCookie);
         }
+        cookieJar.clear();
+        cookieJar.saveFromResponse(null, retainedCookies);
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/support/CouchbaseLiteHttpClientFactory.java]: The first parameter is null.

Comment (hideki) [inline src/main/java/com/couchbase/lite/support/CouchbaseLiteHttpClientFactory.java]: Current implementation of `CookieJar.saveFromResponse(HttpUrl, List<Cookie>)` ignores first `HttpUrl` parameter. I will fix this with [the ticket](https://github.com/couchbase/couchbase-lite-java-core/issues/694).

**Pull request merged**